opnode/rollup/driver/step.go

```diff
@@ -137,52 +159,73 @@ func (d *outputImpl) step(ctx context.Context, l2Head eth.BlockID, l2Finalized e
+	for i, batch := range batches {
+		var txns []l2.Data
+		txns = append(txns, l1InfoTx)
+		if i == 0 {
+			txns = append(txns, deposits...)
+		}
+		txns = append(txns, batch.Transactions...)
+		attrs := &l2.PayloadAttributes{
+			Timestamp:             hexutil.Uint64(batch.Timestamp),
+			Random:                l2.Bytes32(l1Info.MixDigest()),
+			SuggestedFeeRecipient: d.Config.FeeRecipientAddress,
+			Transactions:          txns,
+			NoTxPool:              false,
+		}
+
+		payload, err := d.addBlock(ctx, fc, attrs, true, updateUnsafeHead)
 		if err != nil {
-			return last, fmt.Errorf("failed to extend L2 chain at block %d/%d of epoch %d: %v", i, len(attrsList), epoch, err)
+			return last, fmt.Errorf("failed to extend L2 chain at block %d/%d of epoch %d: %w", i, len(batches), epoch, err)
 		}
 		last = payload.ID()
+		fc.HeadBlockHash = last.Hash
 	}
 
 	return last, nil
 }
 
-// AddBlock extends the L2 chain by deriving the full execution payload from inputs,
-// and then executing and persisting it.
-//
-// After the step completes it returns the block ID of the last processed L2 block, even if an error occurs.
-func AddBlock(ctx context.Context, logger log.Logger, rpc DriverAPI,
-	l2Parent eth.BlockID, l2Finalized common.Hash, attrs *l2.PayloadAttributes) (*l2.ExecutionPayload, error) {
-
-	payload, err := derive.ExecutionPayload(ctx, rpc, l2Parent.Hash, l2Finalized, attrs)
+func (d *outputImpl) addBlock(ctx context.Context, fc l2.ForkchoiceState, attrs *l2.PayloadAttributes, updateSafe, updateUnsafe bool) (*l2.ExecutionPayload, error) {
+	fcRes, err := d.l2.ForkchoiceUpdate(ctx, &fc, attrs)
 	if err != nil {
-		return nil, fmt.Errorf("failed to derive execution payload: %v", err)
+		return nil, fmt.Errorf("failed to create new block via forkchoice: %w", err)
 	}
-
-	logger = logger.New("derived_l2", payload.ID())
-	logger.Info("derived full block", "l2Parent", l2Parent, "attrs", attrs, "payload", payload)
-
-	err = l2.ExecutePayload(ctx, rpc, payload)
+	id := fcRes.PayloadID
+	if id == nil {
+		return nil, errors.New("nil id in forkchoice result when expecting a valid ID")
+	}
+	payload, err := d.l2.GetPayload(ctx, *id)
 	if err != nil {
-		return nil, fmt.Errorf("failed to apply execution payload: %v", err)
+		return nil, fmt.Errorf("failed to get execution payload: %w", err)
 	}
-	logger.Info("executed block")
-
-	err = l2.ForkchoiceUpdate(ctx, rpc, payload.BlockHash, l2Finalized)
+	err = d.l2.ExecutePayload(ctx, payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to insert execution payload: %w", err)
+	}
+	if updateSafe {
+		fc.SafeBlockHash = payload.BlockHash
+	}
+	if updateUnsafe {
+		fc.HeadBlockHash = payload.BlockHash
+	}
+	_, err = d.l2.ForkchoiceUpdate(ctx, &fc, nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to persist execution payload: %v", err)
+		return nil, fmt.Errorf("Could not make the new L2 block canonical via forkchoice: %w", err)
```

Comment:
nit: don't start error messages with a capital, Go style rules: https://github.com/golang/go/wiki/CodeReviewComments#error-strings  (I know, probably the most subjective thing in the whole guide, but my IDE will complain about it every time I open the file)

Reply:
I didn't realize this. Will fix. Maybe we should add a lint for this